test/blackbox-tests/test-cases/lib-collision/lib-collision-public-same-folder.t

```diff
@@ -19,9 +19,9 @@ the same folder.
 Without any consumers of the libraries
 
   $ dune build
-  Error: Library foo is defined twice:
-  - dune:6
-  - dune:3
+  Error: Multiple rules generated for _build/default/foo.cmxs:
```

Comment:
Still pretty blocked 😞 There seems to be a database for all libraries, but then a separate DB for public libs. This one is the one used to resolve public libraries apparently, and it's indexed by the public libs public names.

So when there are only public libraries that conflict, the collision detection won't work because the name resolution will go through the `Project` branch:

https://github.com/ocaml/dune/blob/40ea41fcd935de652b1afaf8ca49f9281a51af31/src/dune_rules/scope.ml#L119-L121

At that point, all the potential duplicates are lost, so the rules creation proceeds and the conflictive cmxs rules are created.

@rgrinberg do you have any pointers? Should we somehow keep 2 indexes in `public_libs`, one keyed by public names and another with private? Thanks.

Reply:
I attempted a fix in https://github.com/ocaml/dune/pull/10179/commits/b970f3c8a2505cac50197171266f4bc70aa42990.

Reply:
The idea is to perform a query to the DB using the libraries private names before adding their rules in `Gen_rules.of_stanza`. Then check the result of this query for `Invalid` values, and proceed accordingly to show the user error.

Reply:
After taking a brief look at the code, I would have thought:

- we create a DB with all the public libraries in the project from the stanzas
    - this DB has `~parent:(Some installed_libraries)`
- when we create the DB with all the public libraries in the project, we could change the map that we generate:
```diff
- Found_or_redirect.t Lib_name.Map.t
+ Found_or_redirect.t list Lib_name.Map.t
```

Given this change, I think you could perform your check like you did for the private libraries?

Reply:
The `Found_or_redirect.t Lib_name.Map.t` type is the one used to create the "private" database, the public database type is `redirect_to Lib_name.Map.t`.

I tried changing the former type as suggested, the result can be seen in https://github.com/jchavarri/dune/pull/11 (draft PR). This change doesn't affect behavior, but I think it makes the pattern matching a bit more noisy.

The main problem, as far as I understand, is that the public DB is indexed by public lib names, but the name collisions happen when the private names are used (i.e. in the private DB). Because public lib installation doesn't query the private DB but the public one, then the conflicts are never detected. That's why I had to add some code that tries to resolve the library private names:

https://github.com/ocaml/dune/blob/531e4d315ea3958cf55884811de80104191d1d41/src/dune_rules/gen_rules.ml#L115-L118

Reply:
I see. The code in `Gen_rules` doesn't match my intuition about this change, but I'm not intimately familiar with `Lib` and `Scope` apart from my cross-compilation changes.

Since you have a significant amount of context about this now, could you summarize the invariants/requirements we're trying to achieve? I'll take a stab at an incomplete list:

- **requirement**: resolving a library by its private name should return all libraries matching that name
    - we filter them based on the context they belong to, expect to find only 1, error if that's not the case
    - **invariant**: resolving a private library returns 1 library per context, or errors out if libs_in_current_context > 1
- **requirement**: resolving a public library by its public name should return all libraries matching that name
    - however, for public libraries, there could exist public libraries `a` and `b` with a common private name `priv`, e.g.


  ```clj
  (library
   (name priv)
   (public_name a)) 
  (library
   (name priv)
   (public_name b))
  ```

    - In this case, we must go look for the private library name to check whether there's another conflicting library (which could be both public or private -- we might need a test with a public library that has the same private name as a private lib?)
    - **invariant**: the transitive closure of libraries for a module group must not contain multiple public libraries with the same private name



I tried to summarize my understanding of the current requirements to help me review the PR better. Did I get anything wrong and/or, is there something I missed?

Reply:
To be honest, I don't think I understand everything that is going on to the level that I can explain what the invariants in too much detail. At the (very) basic level, I could say that:

- before this PR, there was an invariant (implemented in the two DB creation functions) that prevented having two libraries sharing either private names or public names.
- after this PR, we need to relax this invariant to allow two or more libraries with the same private name or public name, because they might exist in different contexts, and at library creation time we can't resolve the `enabled_if` conditions as we don't have the expander available yet.

Maybe one important piece of information (that I recently understood) is that the "redirects" in the private DB are redirecting from the private names to the public names, which is a bit counterintuitive. I would have expected redirects to go from public names to private ones, because all libs have a private name, but not all have a public name. But I assume there are good reasons for that.

This means that the private DB can contain `Found` values with either public names, or private names. This can be seen on the usage of `Library.best_name conf` when `Found` values are added into it:

https://github.com/ocaml/dune/blob/171c23129e4c86ad34bea4750d7fa4608d3019c2/src/dune_rules/scope.ml#L69

As an example of the above, if we have a public library with private name `foo` and public name `bar.foo`, there will be two entries in the private DB map:

- a `Redirect` value, indexed with key `foo`
- a `Found` value, indexed with key `bar.foo`

On cases where the conflict exists on the private names of public libraries (like the one tracked in `lib-collision-public-same-folder.t`), the conflict is detected if we ever have to resolve the private names. But for installing public libs, that's never the case, we will just look for a library called `bar.foo`, get a `Found` value back, and then continue (until we get an error because both libs are trying to install cmxs under the same folder, because for that installation the private names are used).

Reply:
> The code in `Gen_rules` doesn't match my intuition about this change

Maybe the call to `Lib.DB.find_invalid` could be added somewhere else, it doesn't necessarily need to be in `Gen_rules`, I just didn't know where to place it.